### ISSUE TITLE
Warn against non-obvious const shadowing

### ISSUE THREAD

**Issue by FineFindus** [rust-lang/rust-clippy#12959]
### What it does

Constants can be shadowed when in different scopes. This can lead to non-obvious situations, where a later declared constant already shadows the previous constant, e.g. the example below counter-intuitively prints "Hello Rust!" twice. The lint should suggest moving the second constant above the first invocation of it.

### Advantage

- Avoids confusion about which constant is used

### Drawbacks

Such situations may be rare, so a lint may not be necessary.

### Example

```rust
const MESSAGE: &str = "🦀";

fn main() {
    println!("{}", MESSAGE);
    const MESSAGE: &str = "Hello Rust!";
    println!("{}", MESSAGE);
}
```

Could be written as:

```rust
const MESSAGE: &str = "🦀";

fn main() {
    const MESSAGE: &str = "Hello Rust!";
    println!("{}", MESSAGE);
    println!("{}", MESSAGE);
}
```

**Comment by samueltardieu** [comment]
Funny situation indeed.

@rustbot claim

**Comment by lukaslueg** [comment]
The following will print "🦀 Hello Rust! Launch the missiles! 🦀".

```rust
const MESSAGE: &str = "🦀";

fn main() {
    println!("{}", MESSAGE);
    {
        println!("{}", MESSAGE);
        println!("{}", MESSAGE2);
        const MESSAGE2: &str = "Launch the missiles!";
        const MESSAGE: &str = "Hello Rust!";
    }
    println!("{}", MESSAGE);
}
```

I guess op's intention of "non-obvious const shadowing" can be generalized into "const seemingly declared after it is used". If the local const does shadow an outer one, it is particularly bad; but afaics it is always worth linting if the const's value is presumably set _after_ it has already been used; in the example above, `Hello Rust!` in line 6 is spooky action at a distance coming from line 9.
